tq/transfer_queue.go

```diff
@@ -572,7 +572,7 @@ func (q *TransferQueue) enqueueAndCollectRetriesFor(batch batch) (batch, error)
 				}
 			}
 
-			return next, errors.NewRetriableError(err)
+			return next, err
```

Comment:
Hmm, I'm not sure this is correct.  This reverts a change in #3930.  I think we may need to check whether `err` is `nil` here and make an appropriate check so as not to regress that change.

Reply:
I wasn't aware of #3930, but you're right, this would probably introduce a regression. I've added 906890ecb371579ef9f9dd9b5fa268e19c58cc71 which conditionally converts the error, meaning the code-path should pretty much the same with the exception that `nil` is now properly handled. This keeps the required changes and side-effects as low as possible.

On the other hand, I'm not sure if we should improve the _[retry-later code-path](https://github.com/git-lfs/git-lfs/blob/906890ecb371579ef9f9dd9b5fa268e19c58cc71/tq/transfer_queue.go#L568)_ in general. The decision to overwrite the original error should be taken on a batch level and not just based on the first object.

WDYT?

Reply:
Sure, I think that sounds reasonable.  If you want to add a patch for that, that would be very welcome; otherwise, I'll try to find some time to sneak in a patch.

Reply:
@bk2204 I was trying to optimise the code-path for the *retry-later* case a bit and probably stumbled upon another issue. Before I start to tackle that one, I just wanted to make sure my assumptions here are correct.

Setting `err` to `nil` for the `canRetryObjectLater` case during the **first** iteration leads to all subsequent passes being not correctly identified as *"retry-able-later"* and will hence trigger the `else` case.

This has the effect that only the first OID is added to the retry queue and therefore gets successfully uploaded. This will lead to an overall failure, since not all objects have been uploaded:

See the following trace with modified integration test to upload two objects instead of one:

```txt
13:44:57.470911 trace git-lfs: tq: sending batch of size 2                            
#     13:44:57.471118 trace git-lfs: api: batch 2 files
#     13:44:57.471190 trace git-lfs: creds: git credential cache ("http", "127.0.0.1:64702", "batch-upload-retry-later")
#     13:44:57.471195 trace git-lfs: Filled credentials for http://127.0.0.1:64702/batch-upload-retry-later
#     13:44:57.471204 trace git-lfs: HTTP: POST http://127.0.0.1:64702/batch-upload-retry-later.git/info/lfs/objects/batch
#     13:44:57.471734 trace git-lfs: HTTP: 429
#     13:44:57.471756 trace git-lfs: HTTP: rate limit reached
#     13:44:57.471794 trace git-lfs: api error: LFS: Error
#     13:44:57.471808 trace git-lfs: tq: enqueue retry #1 after 11.00s for "6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734" (size: 24)
#     13:45:08.473361 trace git-lfs: tq: sending batch of size 1
#     13:45:08.473540 trace git-lfs: api: batch 1 files
#     13:45:08.473780 trace git-lfs: creds: git credential cache ("http", "127.0.0.1:64702", "batch-upload-retry-later")
#     13:45:08.473793 trace git-lfs: Filled credentials for http://127.0.0.1:64702/batch-upload-retry-later
#     13:45:08.473813 trace git-lfs: HTTP: POST http://127.0.0.1:64702/batch-upload-retry-later.git/info/lfs/objects/batch
#     13:45:08.474822 trace git-lfs: HTTP: 200
#     13:45:08.474875 trace git-lfs: HTTP: {"objects":[{"oid":"6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734","size":24,"actions":{"upload":{"href":"http://127.0.0.1:64702/storage/6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734?r=batch-upload-retry-later","expires_at":"0001-01-01T00:00:00Z"}}}]}
#     13:45:08.474996 trace git-lfs: tq: starting transfer adapter "basic"
13:45:08.475387 trace git-lfs: HTTP: PUT http://127.0.0.1:64702/storage/6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734
#     13:45:08.475898 trace git-lfs: HTTP: 200
Uploading LFS objects:  50% (1/2), 24 B | 0 B/s, done.                                
#     13:45:08.476084 trace git-lfs: filepathfilter: rewrite ".git" as "**/.git/**" (strict: false)
#     13:45:08.476105 trace git-lfs: filepathfilter: rewrite "**/.git" as "**/.git" (strict: false)
#     13:45:08.476178 trace git-lfs: filepathfilter: accepting "tmp"
#     13:45:08.478385 run-command.c:667       trace: run_command: git send-pack --stateless-rpc --helper-status --thin --no-progress http://127.0.0.1:64702/batch-upload-retry-later/ --stdin
#     13:45:08.483675 git.c:455               trace: built-in: git send-pack --stateless-rpc --helper-status --thin --no-progress http://127.0.0.1:64702/batch-upload-retry-later/ --stdin
#     13:45:08.484927 run-command.c:667       trace: run_command: git pack-objects --all-progress-implied --revs --stdout --thin --delta-base-offset -q
#     13:45:08.489089 git.c:455               trace: built-in: git pack-objects --all-progress-implied --revs --stdout --thin --delta-base-offset -q
#     To http://127.0.0.1:64702/batch-upload-retry-later
#      * [new branch]      main -> main
```

Changing the code to determine if all OIDs are *retry-able-later* at the end of the loop should circumvent this. Something along the lines of:

```diff
@@ -557,22 +557,26 @@ func (q *TransferQueue) enqueueAndCollectRetriesFor(batch batch) (batch, error)
                var err error
                bRes, err = Batch(q.manifest, q.direction, q.remote, q.ref, batch.ToTransfers())
                if err != nil {
+                       var isRetriedLater = true
                        // If there was an error making the batch API call, mark all of
                        // the objects for retry, and return them along with the error
                        // that was encountered. If any of the objects couldn't be
                        // retried, they will be marked as failed.
                        for _, t := range batch {
                                if q.canRetryObject(t.Oid, err) {
+                                       isRetriedLater = false
                                        enqueueRetry(t, err, nil)
                                } else if readyTime, canRetry := q.canRetryObjectLater(t.Oid, err); canRetry {
-                                       err = nil
                                        enqueueRetry(t, err, &readyTime)
                                } else {
+                                       isRetriedLater = false
                                        q.wait.Done()
                                }
                        }
 
-                       if err != nil {
+                       if isRetriedLater {
+                               err = nil
+                       } else {
                                err = errors.NewRetriableError(err)
                        }
```

This will lead to the following trace which actually seems correct to me:

```txt
13:49:45.417452 trace git-lfs: tq: sending batch of size 2                            
#     13:49:45.417613 trace git-lfs: api: batch 2 files
#     13:49:45.417678 trace git-lfs: creds: git credential cache ("http", "127.0.0.1:64753", "batch-upload-retry-later")
#     13:49:45.417682 trace git-lfs: Filled credentials for http://127.0.0.1:64753/batch-upload-retry-later
#     13:49:45.417690 trace git-lfs: HTTP: POST http://127.0.0.1:64753/batch-upload-retry-later.git/info/lfs/objects/batch
#     13:49:45.418140 trace git-lfs: HTTP: 429
#     13:49:45.418155 trace git-lfs: HTTP: rate limit reached
#     13:49:45.418178 trace git-lfs: api error: LFS: Error
#     13:49:45.418205 trace git-lfs: tq: enqueue retry #1 after 11.00s for "6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734" (size: 24): batch response: LFS: Error
#     13:49:45.418212 trace git-lfs: tq: enqueue retry #1 after 11.00s for "5404ee773986b403c227c48c61913b32109d8b2aa512b6fba044acc14cfe30a5" (size: 17): batch response: LFS: Error
#     13:49:56.419113 trace git-lfs: tq: sending batch of size 2
#     13:49:56.419316 trace git-lfs: api: batch 2 files
#     13:49:56.419524 trace git-lfs: creds: git credential cache ("http", "127.0.0.1:64753", "batch-upload-retry-later")
#     13:49:56.419537 trace git-lfs: Filled credentials for http://127.0.0.1:64753/batch-upload-retry-later
#     13:49:56.419556 trace git-lfs: HTTP: POST http://127.0.0.1:64753/batch-upload-retry-later.git/info/lfs/objects/batch
#     13:49:56.420483 trace git-lfs: HTTP: 200
#     13:49:56.420504 trace git-lfs: HTTP: {"objects":[{"oid":"6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734","size":24,"actions":{"upload":{"href":"http://127.0.0.1:64753/storage/6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734?r=batch-upload-retry-later","expires_at":"0001-01-01T00:00:00Z"}}},{"oid":"5404ee773986b403c227c48c61913b32109d8b2aa512b6fba044acc14cfe30a5","size":17,"actions":{"upload":{"href":"http://127.0.0.1:64753/storage/5404ee773986b403c227c48c61913b32109d8b2aa512b6fba044acc14cfe30a5?r=batch-upload-
#     13:49:56.420575 trace git-lfs: HTTP: retry-later","expires_at":"0001-01-01T00:00:00Z"}}}]}
13:49:56.420704 trace git-lfs: tq: starting transfer adapter "basic"                  
#     13:49:56.421158 trace git-lfs: HTTP: PUT http://127.0.0.1:64753/storage/6dd8d52c22b4ef5dd1579858ab09263102fbf167e2a708c3a18d1aeb8bb41734
#     13:49:56.421630 trace git-lfs: HTTP: 200
#     13:49:56.421648 trace git-lfs: HTTP: PUT http://127.0.0.1:64753/storage/5404ee773986b403c227c48c61913b32109d8b2aa512b6fba044acc14cfe30a5
#     13:49:56.422100 trace git-lfs: HTTP: 200
Uploading LFS objects: 100% (2/2), 41 B | 0 B/s, done.                                
#     13:49:56.422326 trace git-lfs: filepathfilter: rewrite ".git" as "**/.git/**" (strict: false)
#     13:49:56.422359 trace git-lfs: filepathfilter: rewrite "**/.git" as "**/.git" (strict: false)
#     13:49:56.422431 trace git-lfs: filepathfilter: accepting "tmp"
#     13:49:56.424351 run-command.c:667       trace: run_command: git send-pack --stateless-rpc --helper-status --thin --no-progress http://127.0.0.1:64753/batch-upload-retry-later/ --stdin
#     13:49:56.429435 git.c:455               trace: built-in: git send-pack --stateless-rpc --helper-status --thin --no-progress http://127.0.0.1:64753/batch-upload-retry-later/ --stdin
#     13:49:56.430613 run-command.c:667       trace: run_command: git pack-objects --all-progress-implied --revs --stdout --thin --delta-base-offset -q
#     13:49:56.434771 git.c:455               trace: built-in: git pack-objects --all-progress-implied --revs --stdout --thin --delta-base-offset -q
#     To http://127.0.0.1:64753/batch-upload-retry-later
#      * [new branch]      main -> main
```

Do you see this similar? If so, I can tidy-up above's changes and add them to this MR.

Reply:
Yeah, that definitely seems like the right way forward.  I agree that would be the right thing to do in this case.

Reply:
@bk2204 I've tidied-up the patch and incorporated it into this MR. I've also extended the integration tests to cover a multiple-file scenario in case of a `429 - retry-after` error to prevent regressions further down the line.

The CI pipeline passes in my fork (see https://github.com/fh1ch/git-lfs/actions/runs/1132277274), so this MR would be ready for a first review round from your end.